### PR TITLE
Fix when containerd restarted, event handler may exit

### PR DESCRIPTION
**Pull request by moypray**
**- What I did**

Fix a bug:
Description:
  Kill docker-containerd continuously, and use `kill -SIGUSR1
<dockerpid>` to check docker callstacks. And we will find that event
handler: `startEventsMonitor` or `handleEventStream` will exit.

This will only happen when system is busy, containerd need more time to
startup, and the monitor gorotine maybe exit.

**- How I did it**

**- How to verify it**
kill the containerd process continuously and use `kill -SIGUSR1 <docker>` to see the calltrace.
Will find the goroutine which handle the event loses.

**- Description for the changelog**

Signed-off-by: Wentao Zhang <zhangwentao234@huawei.com>

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by moypray** [comment]
TCs are all passed but failed to kill docker daemon...

**Comment by cpuguy83** [comment]
ping @tonistiigi @mlaventure

File: libcontainerd/remote_unix.go [hunk 1]
```diff
@@ -280,9 +280,21 @@ func (r *remote) startEventsMonitor() error {
...
+		if err == nil {
+			break
+		}
+		logrus.Warnf("libcontainerd: failed to get events from containerd: %q", err)
+
+		if r.closeManually {
+			// ignore error if grpc remote connection is closed manually
+			return nil
+		}
+		<-time.After(1 * time.Second)
```
Comment (mlaventure) [inline libcontainerd/remote_unix.go]: `100 * time.Millisecond`?

File: libcontainerd/remote_unix.go [hunk 2]
```diff
@@ -280,9 +280,21 @@ func (r *remote) startEventsMonitor() error {
 	er := &containerd.EventsRequest{
 		Timestamp: tsp,
 	}
-	events, err := r.apiClient.Events(context.Background(), er, grpc.FailFast(false))
-	if err != nil {
-		return err
+	var (
+		events containerd.API_EventsClient
```
Comment (mlaventure) [inline libcontainerd/remote_unix.go]: nit: `var events containerd.API_EventsClient`

**Review by mlaventure — CHANGES_REQUESTED** [review]

**Comment by moypray** [comment]
ping @cpuguy83 @tonistiigi @mlaventure
Resolved the comments, thanks.

File: libcontainerd/remote_unix.go (hunk 2 above)
Comment (moypray) [inline libcontainerd/remote_unix.go]: hi @mlaventure solved

**Comment by mlaventure** [comment]
On second thought, I would have assumed that with `FailFast` set to false, that would have been taken care automatically by the grpc code. Any idea why it wouldn't @cpuguy83 ?

LGTM otherwise

**Review by mlaventure — APPROVED** [review]

**Comment by moypray** [comment]
ping @cpuguy83 @tonistiigi @cpuguy83 ?

thanks~~

**Comment by tonistiigi** [comment]
Not very familiar with this. Why do we need the extra for loop, and wouldn't it be better to set a timeout to the context if failfast is disabled?

**Comment by mlaventure** [comment]
@tonistiigi if failfast is disabled grpc should keep retrying by itself which is why I don't understand where the issue is from.

> // FailFast configures the action to take when an RPC is attempted on broken
// connections or unreachable servers. If failfast is true, the RPC will fail
// immediately. Otherwise, the RPC client will block the call until a
// connection is available (or the call is canceled or times out) and will retry
// the call if it fails due to a transient error. Please refer to
// https://github.com/grpc/grpc/blob/master/doc/fail_fast.md

**Comment by moypray** [comment]
Hi, I am not very familiar with GRPC too. But in my press testing environment, the GO routine  `handleEventStream ` of docker  disappeared and docker daemon can not receive any container event from containerd. 

That affects a lot, if docker could not handle these events , the containers could not exit or start successfully.

**Comment by tonistiigi** [comment]
@moypray Based on the fix `handleEventStream` should never be called instead of it disappearing.  Do you see what error is logged when `Events` currently fails for you?

**Comment by moypray** [comment]
@tonistiigi Sorry for late, I am busy on other things.
I have seen 2 kinds of errors, I am sure they look like this:
```
error: rpc error: code = 14 desc = grpc: the connection is unavailable"
error: rpc error: code = 13 desc = transport is closing"
```
And today， I found that the goroutine of `handleConnectionChange` will exit too. the error log is 
```
2017-05-03T14:18:30.163173+00:00|info|docker[-]|time="2017-05-03T14:18:30.161475751Z" level=debug msg="libcontainerd: containerd health check returned error: rpc error: code = 14 desc = grpc: the connection is unavailable"
2017-05-03T14:18:38.185628+00:00|info|docker[-]|time="2017-05-03T14:18:38.185044385Z" level=debug msg="libcontainerd: containerd health check returned error: rpc error: code = 13 desc = transport is closing"
```
The second error will make handleConnectionChange exit. I will make a patch for it too. 

Actually, we may meet both of them while recevicing any grpc message, I have looking into the grpc code, I found that there is a goroutine in background to flush the rpc status,  The error types depends on when the connection is flushed ( Am I right? correct me if something wrong)

**Comment by moypray** [comment]
I am sorry, click `Close and comment`,  Reopen it.

**Comment by tonistiigi** [comment]
@moypray I noticed that the error from `startEventsMonitor ` may currently be swallowed. https://github.com/moby/moby/blob/b9ee31ae027bbd62477fea3f58023c90f051db00/libcontainerd/remote_unix.go#L301 And in the case it isn't there isn't a way to distinguish it from other grpc errors. Are you sure these errors appear on startup and from the `Events()` call?

@mlaventure If the `connection is closing` is caused by containerd crashing would there be a log of that.

**Comment by moypray** [comment]
@tonistiigi  I am sure about it, this case is tested by the IV tester in our team, and I use `kill -SIGUSR1 <docker-pid>` to check all the go routines, but doesn't find the backend go routine anymore. and lots of grpc errors were reported in system log.

And I use a script, to kill contianerd continuously per 10 secs, run for about 15 minutes, it could be reproduced.

**Comment by mlaventure** [comment]
a check for `r.closeManually` is missing [here](https://github.com/moby/moby/blob/140fefd581e6ec3cc1e2d434c2f484d2d386331c/libcontainerd/remote_unix.go#L157) (and there's a typo for `closedManually` :sweat_smile:)

The event code handling appears to be correct and reschedule itself. The actual go routine won't show until the connection is made in [here](https://github.com/moby/moby/blob/140fefd581e6ec3cc1e2d434c2f484d2d386331c/libcontainerd/remote_unix.go#L283)

**Comment by moypray** [comment]
@mlaventure  Fix the typo.  Sorry for late replay.

**Comment by moypray** [comment]
@mlaventure  @thaJeztah  I think there are two defects which could make the handler exit.
One is fixed by https://github.com/moby/moby/pull/32986  This PR is another condition.

I should put these two PR together. but I thought that this one is nearly to be merged. so submit #32986

**Comment by mlaventure** [comment]
becasue we're using `FailFast(false)`, I do think the issue shouldn't occur anymore.

**Comment by moypray** [comment]
I am sure that I have the `FailFast(false)` option while this issue occur just like below. but it still has the issue. 
```
events, err = r.apiClient.Events(context.Background(), er, grpc.FailFast(false))
```

@mlaventure  Does the grpc code fix some issue like this??

**Comment by mlaventure** [comment]
@moypray I meant with https://github.com/moby/moby/pull/32986 this should not longer occur because the loop that will end up restarting the goroutine would no longer exit unless the daemon is shutting down.

**Comment by moypray** [comment]
@mlaventure   You have pointed out that [here](https://github.com/moby/moby/blob/140fefd581e6ec3cc1e2d434c2f484d2d386331c/libcontainerd/remote_unix.go#L283)  could make the event handler exit by caller [Line301 ](https://github.com/moby/moby/blob/140fefd581e6ec3cc1e2d434c2f484d2d386331c/libcontainerd/remote_unix.go#L301
) when an grpc error reported.

 #32986 just fixed the health check handler exits by grpc error. it is another go routine.

Sorry I should submit these two patches together, but #32986 is found much later and when I found it I thought this one is nearly to be merged, so submit a new PR #32986 .

And later, I spent a long terrible days to work on other things.  Didn't come here. sorry.

**Comment by mlaventure** [comment]
Hmm, I initially thougth that since the call won't fail if the connection is unavailable or broken it would do.

But I failed to take in account the fact that someone could externally ask `containerd` to shutdown, case which would prompt to return even with ``FailFast(false)`.

**Comment by mlaventure** [comment]
@moypray you missed one renamed it seem

```
libcontainerd/remote_unix.go:157: r.closeManually undefined (type *remote has no field or method closeManually)
```

**Comment by moypray** [comment]
@mlaventure thanks, update done. waiting for CI.

**Review by mlaventure — APPROVED** [review]
LGTM

**Comment by tonistiigi** [comment]
LGTM

**Comment by moypray** [comment]
ping @ehazlett @tonistiigi  :1234:

**Comment by thaJeztah** [comment]
Thanks @moypray !

**Pull request merged**